schnorr_fun/src/frost/chilldkg.rs

```diff
@@ -0,0 +1,1256 @@
+
+            // ⚠ only do mutations after we're sure everything is ok
+            self.inner.add_input(schnorr, from, input.inner)?;
+
+            for (dest, encrypted_share_contrib) in input.encrypted_shares {
+                let agg_encrypted_share = &mut self.agg_encrypted_shares.get_mut(&dest).unwrap().1;
+                *agg_encrypted_share += encrypted_share_contrib;
+            }
+
+            self.encryption_nonces.push(input.encryption_nonce);
+
+            Ok(())
+        }
+
+        /// Which [`Contributor`]s are we missing input from.
+        pub fn missing_from(&self) -> BTreeSet<u32> {
+            self.inner.missing_from()
+        }
+
+        /// Has the coordinator received input from each [`Contributor`].
+        pub fn is_finished(&self) -> bool {
+            self.inner.is_finished()
+        }
+
+        /// Try and finish input aggregation step.
+        ///
+        /// Returns `None` if [`is_finished`] returns `false`.
+        ///
+        /// [`is_finished`]: Self::is_finished
+        pub fn finish(self) -> Option<AggKeygenInput> {
+            let inner = self.inner.finish()?;
+            Some(AggKeygenInput {
+                inner,
+                encrypted_shares: self.agg_encrypted_shares,
+                encryption_nonces: self.encryption_nonces,
+            })
+        }
+    }
+
+    /// Extract our secret share from the `AggKeygenInput`.
+    ///
+    /// This also validates `agg_input`.
+    pub fn receive_share<H, NG>(
+        schnorr: &Schnorr<H, NG>,
+        my_index: PartyIndex,
+        encryption_keypair: &KeyPair,
+        agg_input: &AggKeygenInput,
+    ) -> Result<PairedSecretShare<Normal, Zero>, simplepedpop::ReceiveShareError>
+    where
+        H: Hash32,
+    {
+        let encrypted_share = agg_input
+            .encrypted_shares
+            .get(&my_index)
+            .map(|(_pk, share)| *share)
+            .unwrap_or_default();
+        let share_scalar = decrypt::<H>(
+            my_index,
+            encryption_keypair,
+            &agg_input.encryption_nonces,
+            encrypted_share,
+        );
+        let secret_share = SecretShare {
+            index: my_index,
+            share: share_scalar,
+        };
+        let paired_secret_share =
+            simplepedpop::receive_share(schnorr, &agg_input.inner, secret_share)?;
+
+        Ok(paired_secret_share)
+    }
+
+    fn encrypt<H: Hash32>(
+        encryption_jobs: BTreeMap<PartyIndex, (Point, Scalar<Secret, Zero>)>,
+        multi_nonce_keypair: KeyPair<Normal>,
+    ) -> BTreeMap<PartyIndex, Scalar<Public, Zero>> {
+        encryption_jobs
+            .iter()
+            .map(|(dest, (encryption_key, share))| {
+                let dh_key = g!(multi_nonce_keypair.secret_key() * encryption_key).normalize();
+                let pad = Scalar::from_hash(H::default().add(dh_key).add(encryption_key).add(dest));
+                let payload = s!(pad + share).public();
+                (*dest, payload)
```

Comment:
This doesn't quite match the encryption of `encpedpop` [here](https://github.com/BlockstreamResearch/bip-frost-dkg/blob/master/python/chilldkg_ref/encpedpop.py#L17).

We calulate dh_key the same way.
Their hash is different to ours:
- `Hash( "encpedpop ecdh", dh_key ||  my_pubkey || their_pubkey || context_data )`
- I think their `my_pubkey`/`pubnonce` we have as `multi_nonce_keypair.public_key()`, not `encryption_key`?
- Our hash is missing context data that sounds important? [See this comment:](https://github.com/BlockstreamResearch/bip-frost-dkg/blob/master/python/chilldkg_ref/encpedpop.py#L40) This hashed ElGamal's "crucial feature is to feed the index of the enckey to the hash function"

Reply:
I instead implemented it like in the paper referenced:

![image](https://github.com/user-attachments/assets/fe678460-6927-42cc-88f2-14071077efe3)

1. Yes there a few instances of this. I am just putting in what I think is actually important.
2. `my_pubkey` is  `encryption_key` and `pubnonce` is `multi_nonce_keypair.public_key()`. I'm not sure what the right name. Maybe it shouldn't be called encryption_key because we later transform it into a signing key.
3. The index is dest.

Reply:
That makes sense. I also misread their code, they are computing the hash in the same way.

Reply:
Yes. `pubnonce` is definitely superfluous since it's a function of `dh_key` and `my_pubkey`. There is an approach to these kinds of "what to hash" question which is just hash every single thing without giving a particular justification for it. That's a fine approach but it requires extra care when specifying the thing otherwise you end up with concatenating stuff like "context_data" everywhere and it gets unwieldy.